### DISCUSSION TITLE
Add EU wide Traffic Data via DataEX2

### DISCUSSION THREAD

**Discussion by sauerbread** [organicmaps/organicmaps#2352] [Ideas]
Hi everyone,
so in a quest for providing traffic info for organic maps Ive [mapped out](https://gist.github.com/sauerbread/a865fe549f622c18d7a59a21c50f4955) pretty much all 27 memberstates + Norway, Iceland, UK and New Zealand (_yes unrelated but, idk._). These are either state or state-owned Sites/Apps/APIs which could be borrowed [NewPipe Extractor](https://github.com/TeamNewPipe/NewPipeExtractor) style, but after adding Estonia Ive discovered Datex2. [Datex2](https://www.datex2.eu/) is pretty much a glorified XML Data Exchange Format that is provided through National Access Points (NAP) and soon to be released [NAPCore](http://napcore.eu/). These NAPs provide a variety of Data Sources from state owned or third party sources, including Traffic Data from goverment agencies (TomTom & Here try really hard, but there are free sources right beneath it). So as an app developer we need to "subscribe" to these data sources for an API key. In general the data is free in some cases, e.g. [Swiss](https://opentransportdata.swiss/en/limit-and-costs/), there might be quota limits. In the docs, Datex2 convienently [suggests](https://docs.datex2.eu/developers/) to bind to the data via JAXB. The EU generously provides a [full list](https://transport.ec.europa.eu/transport-themes/intelligent-transport-systems/road/action-plan-and-directive/national-access-points_en) of participating Countries with others to follow. So either we could "gently" scrape the data from the mapped websites or we use Datex2. What are your thoughts?

_TL;DR: Free EU Traffic Data from Official Sources for us to use._

**Comment by sauerbread** [comment]
Also skimming through the docs there is even more data available than just plain traffic data. There is also parking, service information, electromobility infrastructure, traffic regulations and more available.

**Comment by biodranik** [comment]
That looks promising, thanks for your investigation. It is obvious that to properly implement traffic info for our users we need a backend server that will proxy and cache responses to users' requests. In this case, we can stay within allowed limits and reduce expenses.

The implementation details will be hidden on the server, so probably it should be possible to mix different approaches to collect data from different sources.

What do you think? Any ideas or details on the implementation/architecture?
Also, am I right that it is not required to contribute back users' tracks in realtime?

**Reply by sauerbread** [reply]
Yes, you dont need to contribute user tracks. The traffic info is statisticly collected every 5-10 min through vehicle sensors or webcams on highways e.g. [Greece](http://data.nap.gov.gr/dataset/traffic-data-from-vehicle-detection-sensors-in-attiki-odos). So no tracking is needed. I think Datex2 saves us a lot of time, instead of mapping out every single API interface. I think _"scraping"_ just makes sense for countries that obviously dont contribute to Datex2 like [New Zealand](https://www.journeys.nzta.govt.nz/traffic), [USA](https://www.weather.gov/cys/unitedstatesroadconditions) and [Canada](https://www.gov.mb.ca/mit/roadinfo/canadaroads.html).
I dont know which languages y'all are comfortable with so I guess Java / Kotlin as an obvious server im­ple­men­ta­ti­on choice, altough Im not _very_ experienced with it. I could help out with a Go, Nim, Node im­ple­men­ta­ti­on or try my luck with Java.
Also I guess Open Sourcing the Server wouldn't be problematic because all non Datex2 sources are government sites and allow the use of there data e.g. [New Zealand on using their APIs](https://nzta.govt.nz/traffic-and-travel-information/use-our-data/about-the-apis/)

**Reply by biodranik** [reply]
The language/framework choice depends on the architecture and load estimation. If the service is going to be used by millions of OM (and other apps') users, it means something efficient at scaling, like Golang or maybe Nim.
In my opinion, a good open and free project should not use a lot of hardware resources to avoid excess expenses and fees. Everything should be very efficient from the start.

Are there any open source solutions that already tried to use a similar approach?

**Reply by sauerbread** [reply]
There is an [open issue](https://github.com/osmandapp/Osmand/issues/6878) on OsmAnd, but as far as Ive searched on Github OM would be the world first open app to have it. Golang sounds reasonable. 
Architecture wise I would suggest:
- standard `net/http`, `encoding/xml`, `encoding/json` should be performant enough
- [chi](https://github.com/go-chi/chi) for lightweight http routing.
- [redis](https://github.com/go-redis/redis) for caching. Caching Traffic Infos shouldn't be big on ram usage. _If_ this would run on multiple servers, it could use [master/slave replication](https://redis.io/docs/manual/replication/) and save some quota.
- [caddy](https://caddyserver.com/) as a reverse proxy with automatic let's encrypt https and loadbalancing.
- [xgen](https://github.com/xuri/xgen) for auto `struct` generating from `xsd` schemes. (altough just discovered, might even dont need it)
- Maybe a custom `user-agent` string for usage statistics e.g. [`<platform>:<app ID>:<version string>`](https://github.com/reddit-archive/reddit/wiki/API#rules)

Maybe I try implement a prototype on the weekend and see how well this holds up.

**Reply by biodranik** [reply]
There are several questions to work on:
1. How road segments from the real world will be mapped to our OSM data? E.g. client sends some points (a route), they should be mapped to real roads (OSM data could be wrong or not precise).
2. What should be in the client's request?
3. What should be in the server's response to the client? There is already a (commented out) implementation of some third-party traffic providers in Organic Maps, it should be investigated and taken into an account to plan any client-related changes in the routing algorithm.
4. How would you estimate the RAM/CPU/network load for 1K Qps, 10K Qps, 100K Qps?

**Reply by TobiPeterG** [reply]
Hi there, I would be willing to help, has there been any progress on this issue so far?

**Reply by biodranik** [reply]
@TobiPeterG the progress starts when someone is involved :)

1. Is there any real data already available, so we can start analyzing it?
2. Did you read other related issues? https://github.com/organicmaps/organicmaps/issues/7155
3. There is already a client-side implementation in OM to highlight and use traffic data. So the goal could be to design a server that converts traffic (and other) data into a proper cacheable format for clients.

**Reply by TobiPeterG** [reply]
@biodranik Thank you for your response :)
The only easy accessible source I found is from TomTom and their API is paid and proprietary.
I also registered at mobilithek.info since I'm German and I would like to test it, but the available sources are all only for specific regions and it is not easy to tell how often they get updated.
Maybe we should create our own service, where organisations can upload their traffic data to and create a hub on our  own. We could then feed it with other sources as well.
France seems to have a centralized data source for example.
We could then use it in organic maps and also allow users to report traffic jams (or opt-in for automatic reporting?). We could also try to get other projects liks OSMand to use it and we could build an open service for traffic data :)

I guess a first step would be to create a prototype server that reports traffic jams everywhere or at one specific place to check if it works.

What do you think?

EDIT: Maybe we could build on top of this?
https://github.com/opentraffic/otv2-platform

**Reply by biodranik** [reply]
...organisations discover OM and contribute there traffic info regularly.


Sounds like a dream. It makes sense to focus on listing and retrieving data from existing sources, and converting it into client-friendly format. Adding there data from clients can be done as the next step. 

So yes, building a server side part. Or finding and reusing something that works and fits well.

**Reply by TobiPeterG** [reply]
I've created an initial README summarizing the previous discussions
https://github.com/TobiPeterG/OSMTraffic

Please have a look and check if the points are valid and if I should change or add something :)

**Reply by TobiPeterG** [reply]
I created an initial server in go that can be built with docker-compose:
https://github.com/TobiPeterG/OSMTraffic-server

It uses the Autobahn GmbH APi to get traffic information in Germany and their response is reformated to GEOJson.
Please tell me if this is the desired format that OM could use and how OM would request the traffic info (e.g. coordinates?)

**Reply by dryusername** [reply]
Hi, the Swedish public transport agencies have a project that is related: https://www.trafiklab.se/

They also collaborate with other EU countries to make a universal and open source traffic system with live positioning of all EU public transit vehicles in real time.

Would be great if it could be implemented into the app 😊

**Comment by TobiPeterG** [comment]
I started working on the server side, however, this will a) still take quite some time and b) will not cover a lot, especially in the beginning.
A thought I had today was that maybe it would be useful to implement the TomTom traffic API for now and offer it in a "subscription" version of the app, with the only difference being access to live traffic data. This way, we can already make the app ready for receiving and reporting traffic data, without the open traffic Server being ready and maybe have them in parallel later on, with the open traffic being available to everyone. The subscription could be e.g. 2€/month to cover the cost and support the development of the traffic API. This way, the app would get way more useful for car navigation by having live traffic data and the development of an open alternative would still be supported. :)

Another unrelated thought would be to allow users to have a "bounty" program to fix bugs/implement features, e.g. offering 20€ for fixing bug x, implementing feature y. This would, of course, be optional, but might attract new contributors and motivate overall.
What are your thoughts on this?

**Comment by mule632** [comment]
I've only discovered OM in the last 6 months. I'm interested in helping in its development with donations. Though as discussed here, traffic data (plus eco routing) would really help me fully leave Google maps.

I've had a thought but unsure how realistic. I see there's two optionsbeing talked about. TomTomAPI (quick turn aroubs to get set) and some open source data eerrrr source. I see the reasoning for the TomTomAPI route. I just wondered, how realistic could it be to, say, self host the service that would run the connection to the api (in a docker) and then in organic maps have a setting to point it at it to get the traffic data?
